cpp/src/endpoint.cpp

```diff
@@ -58,8 +58,18 @@ Endpoint::Endpoint(std::shared_ptr<Component> workerOrListener,
   params->err_handler.cb  = Endpoint::errorCallback;
   params->err_handler.arg = _callbackData.get();
 
-  utils::ucsErrorThrow(ucp_ep_create(worker->getHandle(), params, &_handle));
-  ucxx_trace("Endpoint created: %p", _handle);
+  if (worker->isProgressThreadRunning()) {
+    volatile ucs_status_t status = UCS_INPROGRESS;
```

Comment:
I think you want a condition variable, you need a shared mutex, and then something like:
```c++
std::condition_variable cv;
std::mutex m;
ucs_status_t status = UCS_INPROGRESS;
worker->registerGenericPre([this, params, &status, &cv, &m]() {
    auto worker = ...;
    {
    std::lock_guard lock(m);
    status = ...;
    }
    cv.notify_one();
});
{
std::unique_lock lock(m);
cv.wait(lock, [&status]() { return status != UCS_INPROGRESS; });
}
```

(untested, ...);

Reply:
OK, but `volatile` only works like that in single-threaded programs. It has nothing to say about reading between threads (indeed, as referenced above, reading a volatile variable without synchronisation is UB).

Reply:
How so? How would the same thread modify it in some other way? `volatile` is often used by low-level applications that rely on "external causes" to modify it (such as an interrupt), a thread would also be an "external cause" that modifies the memory of `status` and thus requires re-reading is done in each iteration.

From this [SO post](https://stackoverflow.com/a/2485177):

> For thread-safe accesses to shared data, we need a guarantee that:
>
> * the read/write actually happens (that the compiler won't just store the value in a register instead and defer updating main memory until much later)
> * that no reordering takes place. Assume that we use a volatile variable as a flag to indicate whether or not some data is ready to be read. In our code, we simply set the flag after preparing the data, so all looks fine. But what if the instructions are reordered so the flag is set first?
> volatile does guarantee the first point. It also guarantees that no reordering occurs between different volatile reads/writes. All volatile memory accesses will occur in the order in which they're specified. That is all we need for what volatile is intended for: manipulating I/O registers or memory-mapped hardware, but it doesn't help us in multithreaded code where the volatile object is often only used to synchronize access to non-volatile data. Those accesses can still be reordered relative to the volatile ones.

For the use case here we're interested in the first point, which is reported to be guaranteed. The synchronization problem is indeed not guaranteed, but it isn't a problem in here since we'll have no concurrency on writing and we don't need to ensure order which `status` is ultimately read (i.e., it will work whether its ready for the first time before or after the callback writes to it).

Reply:
That's not sufficient (or at least not portable) because you have no guarantee that writes/reads are atomic (for example), so you might have torn reads, depending on the size of the `ucs_status_t` object and the hardware you're running on. mutex + condition_variable is AIUI "the right way" to handle this and should probably be at least as well-performing (possibly more if the worker progress thread takes a long time).

Reply:
Thinking about this, and rewatching from 1:04 here https://youtu.be/KeLBd2EJLOU?t=3889, I think this can also be done with atomics, which is probably the best bet:

```c++
std::atomic<ucs_status_t> status = UCS_INPROGRESS;
worker->registerGenericPre([this, params, &status]() {
    auto worker = ...;
    status.store(ucp_ep_create(...), std::memory_order_release);
});
while(status.load(std::memory_order_acquire) == UCS_INPROGRESS)
  ;
```

Reply:
This compiles down to a simple [spin](https://gcc.godbolt.org/z/oooW17anq) if the type we're writing is small enough that the hardware supports atomic reads/writes through the hardware's coherency model
```asm
.L20:
        mov     eax, DWORD PTR [rsp+20]
        test    eax, eax
        je      .L20
```

And emits the correct builtin `__atomic_store`/`__atomic_load` pairs [if not](https://gcc.godbolt.org/z/d3hq6GaWd):

```asm
.L20:
        mov     ecx, 2
        lea     rdx, [rsp+4]
        lea     rsi, [rsp+28]
        mov     edi, 12
        call    __atomic_load
        cmp     DWORD PTR [rsp+8], 0
        je      .L20
```

Reply:
We don't care about portability right now, UCX only supports Linux with a handful of architectures. [`ucs_status_t` is an `enum`](https://github.com/openucx/ucx/blob/4057ee980fc9fb15a8f484ea6e9b485bb74306c8/src/ucs/type/status.h#L45-L86) and fits in a single-register, so that should be no problem either. In the interest of leaving no loose ends here, I've switched to use `condition_variable`s, see https://github.com/rapidsai/ucxx/pull/65/commits/550b1fb3e3454f53f6ab1b7fc2edfc4e215baa82 and https://github.com/rapidsai/ucxx/pull/65/commits/9687d12e2c9c91940f28851c7553dd3c366ae7dc .

Reply:
I hate GH. I didn't see your comment before. Atomics would also work but we would still need to keep on looping. With a `condition_variable` we can let the HW optimizations truly idle the main thread until the `condition_variable` is awaken, performance-wise that would be theoretically better, although, as you can see, code is much longer now.